shoop/admin/modules/orders/json_order_creator.py

```diff
@@ -121,27 +136,93 @@ def _process_lines(self, source, state):
             except Exception as exc:  # pragma: no cover
                 self.add_error(exc)
 
-    def _initialize_source_from_state(self, state, creator):
-        customer_data = state.pop("customer", None) or {}
-        shop_data = state.pop("shop", None) or {}
-        methods_data = state.pop("methods", None) or {}
-        customer = self.safe_get_first(Contact, pk=customer_data.get("id"))
-        shop = self.safe_get_first(Shop, pk=shop_data.pop("id", None))
+    def _create_contact_from_address(self, billing_address, is_company):
+        name = billing_address.get("name", None)
+        if not name:
+            self.add_error(
+                ValidationError(_("Name is required for customer"), code="no_name")
+            )
 
-        if not customer:
-            self.add_error(ValidationError(_("Please choose a valid customer."), code="no_customer"))
+        phone = billing_address.get("phone", "")
+        email = billing_address.get("email", "")
+        fields = {"name": name, "phone": phone, "email": email}
+        if is_company:
+            tax_number = billing_address.pop("tax_number", None)
+            if not tax_number:
+                self.add_error(
+                    ValidationError(_("Tax number is not set for company."), code="no_tax_number")
+                )
+                return
+
+            fields.update({"tax_number": tax_number})
+            customer = CompanyContact.objects.create(**fields)
+        else:
+            customer = PersonContact.objects.create(**fields)
+        return customer
 
+    def _create_address(self, address):
+        address_fields = [
+            "name", "tax_number", "phone", "email", "street", "street2", "postal_code", "city", "region", "country"
+        ]
+        fields = {}
+        for field in address_fields:
+            fields[field] = address.get(field, "")
+
+        address = Address.objects.filter(**fields).first()
```

Comment:
Having addressed denormalized to fields (i.e. order having shipping_street, shipping_zip etc.) is not very nice approach IMHO. What if we want to add a new field to address? We would have to add it to all places having addresses. Also setting/getting addresses would be much harder and error prone, since each field would have to be set/get separately.

I see that there is five options:
* (1) Make address objects used in orders immutable.
* (2) Make address a custom data type and implement a Django model field which will serialize the address to single db column on save and deserializes it back to Address object on load.
  - Would make querying orders by some address field much harder.
* (3) Have every order have its separate address instance.
  - Would "flood" the database as @Pikkupomo said.
* (4) Save addresses in denormalized form to orders (i.e. order having shipping_street, shipping_zip etc.)
  - This would flood the database exactly as much as (3) btw.
* (5) Just ignore it: Use same address ids if fields occasionally match and let addresses change on orders if they are changed elsewhere.
 - Not really an option IMHO.

I would prefer (1). For order creating it would mean that it should pick only immutable addresses or create a new immutable address if no such address exists. We already have immutability of Addresses, but IMHO it should be revised now: Possibly switching it to use ChangeProtected mixin like Shop and Tax do.